Sources/SwiftSyntaxMacrosTestSupportFrameworkAgnostic/Assertions.swift

```diff
@@ -0,0 +1,599 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=6.0)
+import SwiftBasicFormat
+public import SwiftDiagnostics
+@_spi(FixItApplier) import SwiftIDEUtils
+import SwiftParser
+import SwiftParserDiagnostics
+public import SwiftSyntax
+public import SwiftSyntaxMacroExpansion
+private import SwiftSyntaxMacros
+private import _SwiftSyntaxTestSupportFrameworkAgnostic
+#else
+import SwiftBasicFormat
+import SwiftDiagnostics
+@_spi(FixItApplier) import SwiftIDEUtils
+import SwiftParser
+import SwiftParserDiagnostics
+import SwiftSyntax
+import SwiftSyntaxMacroExpansion
+import SwiftSyntaxMacros
+import _SwiftSyntaxTestSupportFrameworkAgnostic
+#endif
+
+/// Defines the location at which the a test failure should be anchored. This is typically the location where the
+/// assertion function is called.
+public struct TestFailureLocation {
+  public let fileID: StaticString
+  public let filePath: StaticString
+  public let line: UInt
+  public let column: UInt
+
+  public init(
+    fileID: StaticString,
+    filePath: StaticString,
+    line: UInt,
+    column: UInt
+  ) {
+    self.fileID = fileID
+    self.filePath = filePath
+    self.line = line
+    self.column = column
+  }
+
+  fileprivate init(underlying: _SwiftSyntaxTestSupportFrameworkAgnostic.TestFailureLocation) {
+    self.init(
+      fileID: underlying.fileID,
+      filePath: underlying.filePath,
+      line: underlying.line,
+      column: underlying.column
+    )
+  }
+
+  /// This type is intentionally different to `_SwiftSyntaxTestSupportFrameworkAgnostic.TestFailureLocation` so we can
+  /// import `_SwiftSyntaxTestSupportFrameworkAgnostic` privately and don't expose its internal types.
+  fileprivate var underlying: _SwiftSyntaxTestSupportFrameworkAgnostic.TestFailureLocation {
+    _SwiftSyntaxTestSupportFrameworkAgnostic.TestFailureLocation(
+      fileID: self.fileID,
+      filePath: self.filePath,
+      line: self.line,
+      column: self.column
+    )
+  }
+}
+
+/// Defines the details of a test failure, consisting of a message and the location at which the l
+public struct TestFailureSpec {
```

Comment:
Have you consider throwing this as an `Error` and `catch let error as TestFailureSpec {`?

Reply:
The problem is that we could only throw a single failure and not multiple.